kv/kvserver/tenantrate: TestDataDriven failed

[(kv/kvserver/tenantrate).TestDataDriven failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2229783&tab=buildLog) on [master@2e4df338ecd027f36a4ea9fee78b7398a5f55271](https://github.com/cockroachdb/cockroach/commits/2e4df338ecd027f36a4ea9fee78b7398a5f55271):

```
=== RUN   TestDataDriven
--- FAIL: TestDataDriven (0.01s)
=== RUN   TestDataDriven/basic
    --- FAIL: TestDataDriven/basic (0.00s)
        limiter_test.go:68: output didn't match expected:
            @@ -1,6 +1,6 @@
            -kv_tenant_rate_limit_current_blocked 2
            -kv_tenant_rate_limit_current_blocked{tenant_id="2"} 2
            +kv_tenant_rate_limit_current_blocked 1
            +kv_tenant_rate_limit_current_blocked{tenant_id="2"} 1
             kv_tenant_rate_limit_current_blocked{tenant_id="3"} 0
             kv_tenant_rate_limit_current_blocked{tenant_id="5"} 0
             kv_tenant_rate_limit_current_blocked{tenant_id="system"} 0
             

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestDataDriven PKG=./pkg/kv/kvserver/tenantrate TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDataDriven.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver/tenantrate).TestDataDriven failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2229784&tab=buildLog) on [master@2e4df338ecd027f36a4ea9fee78b7398a5f55271](https://github.com/cockroachdb/cockroach/commits/2e4df338ecd027f36a4ea9fee78b7398a5f55271):

```
=== RUN   TestDataDriven
--- FAIL: TestDataDriven (0.02s)
=== RUN   TestDataDriven/basic
    --- FAIL: TestDataDriven/basic (0.01s)
        limiter_test.go:68: output didn't match expected:
            @@ -1,6 +1,6 @@
            -kv_tenant_rate_limit_current_blocked 2
            +kv_tenant_rate_limit_current_blocked 1
             kv_tenant_rate_limit_current_blocked{tenant_id="2"} 2
             kv_tenant_rate_limit_current_blocked{tenant_id="3"} 0
             kv_tenant_rate_limit_current_blocked{tenant_id="5"} 0
             kv_tenant_rate_limit_current_blocked{tenant_id="system"} 0
             

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-race -parallel=2

```
make stressrace TESTS=TestDataDriven PKG=./pkg/kv/kvserver/tenantrate TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDataDriven.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver/tenantrate).TestDataDriven failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2231505&tab=buildLog) on [master@8b91062f9351d18f9104aff567cb152df162021e](https://github.com/cockroachdb/cockroach/commits/8b91062f9351d18f9104aff567cb152df162021e):

```
=== RUN   TestDataDriven
--- FAIL: TestDataDriven (0.01s)
=== RUN   TestDataDriven/basic
    --- FAIL: TestDataDriven/basic (0.01s)
        limiter_test.go:68: output didn't match expected:
            @@ -1,6 +1,6 @@
            -kv_tenant_rate_limit_current_blocked 2
            +kv_tenant_rate_limit_current_blocked 1
             kv_tenant_rate_limit_current_blocked{tenant_id="2"} 2
             kv_tenant_rate_limit_current_blocked{tenant_id="3"} 0
             kv_tenant_rate_limit_current_blocked{tenant_id="5"} 0
             kv_tenant_rate_limit_current_blocked{tenant_id="system"} 0
             

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestDataDriven PKG=./pkg/kv/kvserver/tenantrate TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDataDriven.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver/tenantrate).TestDataDriven failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2232684&tab=buildLog) on [master@8746e7e347dbbc19be051168225cd4fab0582294](https://github.com/cockroachdb/cockroach/commits/8746e7e347dbbc19be051168225cd4fab0582294):

```
=== RUN   TestDataDriven
--- FAIL: TestDataDriven (0.01s)
=== RUN   TestDataDriven/basic
    --- FAIL: TestDataDriven/basic (0.00s)
        limiter_test.go:68: output didn't match expected:
            @@ -1,6 +1,6 @@
            -kv_tenant_rate_limit_current_blocked 2
            -kv_tenant_rate_limit_current_blocked{tenant_id="2"} 2
            +kv_tenant_rate_limit_current_blocked 1
            +kv_tenant_rate_limit_current_blocked{tenant_id="2"} 1
             kv_tenant_rate_limit_current_blocked{tenant_id="3"} 0
             kv_tenant_rate_limit_current_blocked{tenant_id="5"} 0
             kv_tenant_rate_limit_current_blocked{tenant_id="system"} 0
             

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestDataDriven PKG=./pkg/kv/kvserver/tenantrate TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDataDriven.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

[(kv/kvserver/tenantrate).TestDataDriven failed](https://teamcity.cockroachdb.com/viewLog.html?buildId=2233905&tab=buildLog) on [master@cd8267ca84f6a0267a898220cfaa6630188b45c5](https://github.com/cockroachdb/cockroach/commits/cd8267ca84f6a0267a898220cfaa6630188b45c5):

```
=== RUN   TestDataDriven
--- FAIL: TestDataDriven (0.01s)
=== RUN   TestDataDriven/basic
    --- FAIL: TestDataDriven/basic (0.00s)
        limiter_test.go:68: output didn't match expected:
            @@ -1,6 +1,6 @@
            -kv_tenant_rate_limit_current_blocked 2
            -kv_tenant_rate_limit_current_blocked{tenant_id="2"} 2
            +kv_tenant_rate_limit_current_blocked 1
            +kv_tenant_rate_limit_current_blocked{tenant_id="2"} 1
             kv_tenant_rate_limit_current_blocked{tenant_id="3"} 0
             kv_tenant_rate_limit_current_blocked{tenant_id="5"} 0
             kv_tenant_rate_limit_current_blocked{tenant_id="system"} 0
             

```

<details><summary>More</summary><p>
Parameters:

- TAGS=
- GOFLAGS=-parallel=4

```
make stressrace TESTS=TestDataDriven PKG=./pkg/kv/kvserver/tenantrate TESTTIMEOUT=5m STRESSFLAGS='-timeout 5m' 2>&1
```


[See this test on roachdash](https://roachdash.crdb.dev/?filter=status%3Aopen+t%3A.%2ATestDataDriven.%2A&sort=title&restgroup=false&display=lastcommented+project)
<sub>powered by [pkg/cmd/internal/issues](https://github.com/cockroachdb/cockroach/tree/master/pkg/cmd/internal/issues)</sub></p></details>

--- a/pkg/kv/kvserver/tenantrate/limiter_test.go
+++ b/pkg/kv/kvserver/tenantrate/limiter_test.go
@@ -327,7 +327,8 @@ func (ts *testState) recordRead(t *testing.T, d *datadriven.TestData) string {
 
 // metrics will print out the prometheus metric values. The command takes an
 // argument as a regular expression over the values. The metrics are printed in
-// lexicographical order.
+// lexicographical order. The command will retry until the output matches to
+// make it more robust to races in metric recording.
 //
 // For example:
 //
@@ -358,6 +359,20 @@ func (ts *testState) recordRead(t *testing.T, d *datadriven.TestData) string {
 //  kv_tenant_rate_limit_write_bytes_admitted{tenant_id="2"} 50
 //
 func (ts *testState) metrics(t *testing.T, d *datadriven.TestData) string {
+	exp := strings.TrimSpace(d.Expected)
+	if err := testutils.SucceedsSoonError(func() error {
+		got := ts.getMetricsText(t, d)
+		if got != exp {
+			return errors.Errorf("got: %q, exp: %q", got, exp)
+		}
+		return nil
+	}); err != nil {
+		d.Fatalf(t, "failed to find expected timers: %v", err)
+	}
+	return d.Expected
+}
+
+func (ts *testState) getMetricsText(t *testing.T, d *datadriven.TestData) string {
 	ex := metric.MakePrometheusExporter()
 	ex.ScrapeRegistry(ts.m, true /* includeChildMetrics */)
 	var in bytes.Buffer
@@ -381,7 +396,8 @@ func (ts *testState) metrics(t *testing.T, d *datadriven.TestData) string {
 		d.Fatalf(t, "failed to process metrics: %v", err)
 	}
 	sort.Strings(outLines)
-	return strings.Join(outLines, "\n")
+	metricsText := strings.Join(outLines, "\n")
+	return metricsText
 }
 
 // timers waits for the set of open timers to match the expected output.

